versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/VersionStoreImpl.java

```diff
@@ -591,20 +594,32 @@ public PaginationIterator<KeyEntry> getKeys(
             ContentKey key = storeKeyToKey(indexElement.key());
             CommitOp commitOp = indexElement.content();
             Content.Type contentType = contentTypeForPayload(commitOp.payload());
+            Content content = null;
 
-            if (withContent) {
-              Content c =
+            UUID contentId = commitOp.contentId();
+            String contentIdString;
+            if (contentId != null) {
+              contentIdString = contentId.toString();
+            } else {
+              // this should only be hit by imported legacy nessie repos
+              content =
                   contentMapping.fetchContent(
```

Comment:
This one can probably be optimized by using bulk-fetching.
The current implementation is quite naive and fetches each `Content` individually - it would save quite some time and DB load when multiple `Content` objects are fetched at once using a bulk-operation. Requires some refactoring to have a 2-stage `PaginationIterator`.

Reply:
yeah i had discussed the same with @dimas-b yesterday but due to the larger refactoring we first fix the unnecessary work and can look at doing bulk loading in a follow-up